source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1024,9 +1024,9 @@ void ClusterManagerImpl::updateClusterCounts() {
   if (all_clusters_initialized && ads_mux_) {
     const auto type_url = Config::getTypeUrl<envoy::config::cluster::v3::Cluster>();
     if (last_recorded_warming_clusters_count_ == 0 && !warming_clusters_.empty()) {
+      resume_cds_.reset();
       resume_cds_ = ads_mux_->pause(type_url);
     } else if (last_recorded_warming_clusters_count_ > 0 && warming_clusters_.empty()) {
-      ASSERT(resume_cds_ != nullptr);
```

Comment:
I wonder if the fix should be to change the statement in the line above to:
```
else if (resume_cds_ != nullptr && warming_clusters_.empty()) {
```

And more generally, whether instead of using `last_recorded_warming_clusters_count_` the value of `resume_cds_` should be used directly?

Reply:
that's a good point, I actually forgot what resume_cds_ did, it basically reduces the pause count and resumes queuing requests when the count is 0.  In that case, we only need one active.  However, the cleanup occurs in the ScopedResume's destructor, which means on L1027, if we just reassign the resume_cds_, it'll invoke the destructor of the previous pointer first, so we'd end up decrementing the pause count twice.  

So maybe what we want is:
```
if (resume_cds_ != nullptr && !warming_clusters_.empty()) {
  resume_cds_ = ads_mux_->pause(type_url);
} else if (warming_clusters_.empty()) {
  resume_cds_.reset();
}
```

I made this change and ran the ads_integration_test suite (which has lots of pause and resumption tests), and they all passed.